File: backend/src/routes/tts.js
const express = require('express');
const router = express.Router();
const ttsService = require('../services/ttsService');

/**
 * 获取可用语音列表
 * GET /api/tts/voices
 */
router.get('/voices', async (req, res) => {
    try {
        const voices = await ttsService.getVoices();
        res.json({
            success: true,
            data: voices
        });
    } catch (error) {
        console.error('获取语音列表失败:', error);
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

/**
 * 合成语音
 * POST /api/tts/synthesize
 * Body: { text, voice, rate, pitch, volume }
 */
router.post('/synthesize', async (req, res) => {
    try {
        const { text, voice, rate, pitch, volume } = req.body;

        if (!text) {
            return res.status(400).json({
                success: false,
                error: '缺少 text 参数'
            });
        }

        const audioData = await ttsService.synthesizeEdge(text, {
            voice,
            rate: rate || 0,
            pitch: pitch || 0,
            volume: volume || 0
        });

        // 返回音频数据
        res.set({
            'Content-Type': 'audio/mpeg',
            'Content-Length': audioData.length,
            'Cache-Control': 'public, max-age=3600'
        });
        res.send(audioData);
    } catch (error) {
        console.error('合成语音失败:', error);
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

/**
 * 流式合成语音
 * POST /api/tts/synthesize-stream
 * Body: { text, voice, rate, pitch, volume }
 */
router.post('/synthesize-stream', async (req, res) => {
    try {
        const { text, voice, rate, pitch, volume } = req.body;

        if (!text) {
            return res.status(400).json({
                success: false,
                error: '缺少 text 参数'
            });
        }

        res.set({
            'Content-Type': 'audio/mpeg',
            'Transfer-Encoding': 'chunked',
            'Cache-Control': 'no-cache'
        });

        await ttsService.synthesizeStream(
            text,
            { voice, rate: rate || 0, pitch: pitch || 0, volume: volume || 0 },
            (chunk) => {
                res.write(chunk);
            }
        );

        res.end();
    } catch (error) {
        console.error('流式合成失败:', error);
        if (!res.headersSent) {
            res.status(500).json({
                success: false,
                error: error.message
            });
        }
    }
});

/**
 * 清理缓存
 * DELETE /api/tts/cache
 */
router.delete('/cache', async (req, res) => {
    try {
        await ttsService.clearCache();
        res.json({
            success: true,
            message: '缓存已清理'
        });
    } catch (error) {
        console.error('清理缓存失败:', error);
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

module.exports = router;

File: python-backend/app/routes/tts.py
"""
TTS (Text-to-Speech) API 路由
"""
from fastapi import APIRouter, Response
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional
from loguru import logger

from app.services.tts_service import TTSService

router = APIRouter()
tts_service = TTSService()


class SynthesizeRequest(BaseModel):
    """语音合成请求"""
    text: str
    voice: Optional[str] = "zh-CN-XiaoxiaoNeural"
    rate: Optional[int] = 0
    pitch: Optional[int] = 0
    volume: Optional[int] = 0


@router.get("/voices")
async def get_voices():
    """
    获取可用语音列表
    GET /api/tts/voices
    """
    try:
        voices = await tts_service.get_voices()
        return {
            "success": True,
            "data": voices
        }
    except Exception as e:
        logger.error(f"获取语音列表失败: {e}")
        return {
            "success": False,
            "error": str(e)
        }


@router.post("/synthesize")
async def synthesize(request: SynthesizeRequest):
    """
    合成语音
    POST /api/tts/synthesize
    Body: { text, voice, rate, pitch, volume }
    """
    try:
        if not request.text:
            return {
                "success": False,
                "error": "缺少 text 参数"
            }
        
        audio_data = await tts_service.synthesize(
            request.text,
            request.voice,
            request.rate,
            request.pitch,
            request.volume
        )
        
        # 返回音频数据
        return Response(
            content=audio_data,
            media_type="audio/mpeg",
            headers={
                "Content-Length": str(len(audio_data)),
                "Cache-Control": "public, max-age=3600"
            }
        )
    except Exception as e:
        logger.error(f"合成语音失败: {e}")
        return {
            "success": False,
            "error": str(e)
        }


@router.post("/synthesize-stream")
async def synthesize_stream(request: SynthesizeRequest):
    """
    流式合成语音
    POST /api/tts/synthesize-stream
    Body: { text, voice, rate, pitch, volume }
    """
    try:
        if not request.text:
            return {
                "success": False,
                "error": "缺少 text 参数"
            }
        
        async def generate():
            """生成音频流"""
            chunks = []
            
            def callback(chunk):
                chunks.append(chunk)
            
            await tts_service.synthesize_stream(
                request.text,
                request.voice,
                request.rate,
                request.pitch,
                request.volume,
                callback
            )
            
            for chunk in chunks:
                yield chunk
        
        return StreamingResponse(
            generate(),
            media_type="audio/mpeg",
            headers={
                "Transfer-Encoding": "chunked",
                "Cache-Control": "no-cache"
            }
        )
    except Exception as e:
        logger.error(f"流式合成失败: {e}")
        return {
            "success": False,
            "error": str(e)
        }


@router.delete("/cache")
async def clear_cache():
    """
    清理缓存
    DELETE /api/tts/cache
    """
    try:
        await tts_service.clear_cache()
        return {
            "success": True,
            "message": "缓存已清理"
        }
    except Exception as e:
        logger.error(f"清理缓存失败: {e}")
        return {
            "success": False,
            "error": str(e)
        }
